### DISCUSSION TITLE
Auxiliary variables

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#21955] [Q&A General] [answered]
Hello buddy,

I have a questions regarding the auxiliary variables.

I see the following words from the workshop "Auxiliary variables come in two flavors: Nodal and Elemental" and at the same time in the auxiliary kernel we wrote velocity expression in this way "-(_permeability[_qp] / _viscosity[_qp]) * _pressure_gradient[_qp]" where qp is the quadrature point. 

It is difficult for me to link this quadrature point with Nodal and Elemental here. Could you please give an example that can intuitively illustrate the odal varaible and elemental variable?
Thank you in advance.

Kind regards,
Qiao

**Comment by GiudGiud** [comment, marked as answer]
Hello

Elemental variables may have multiple quadrature points (CONSTANT MONOMIAL or cell-centered finite volume only have one). _qp indexing reaches the value computed for the relevant quadrature point. Auxkernels for elemental variables loop over elements, then have an inner loop on quadrature points. 

Nodal variables use the nodes as quadrature points. So _qp is just 0

Re-using the same syntax for both allows us to write a single auxkernel definition that is valid for both nodal and elemental variables. 

Guillaume

**Reply by Oops-Qiao** [reply]
ok, one more question here, if we have a variable: "velocity", this velocity can be set as either elemental variable or nodal variable, right? or maybe it can only be nodal variable? 
Another thing is for the _qp in the elemental variable case, it is numbered for each element for example, _qp =1,2,3. or _qp is numbered globally for all the elements?  Thank you again.

**Reply by GiudGiud** [reply]
it could be either. In FE we use nodal variables for velocity, in FV we use an elemental one. The type of variable may influence what discretization you use of the equations (CG, DG etc)

it s numbered locally

**Reply by Oops-Qiao** [reply]
> it could be either. In FE we use nodal variables for velocity, in FV we use an elemental one. The type of variable may influence what discretization you use of the equations (CG, DG etc)
> 
> it s numbered locally

Hi Guillaume,

Here you mentioned that velocity is nodal variable in FE. Do you mean we compute velocity in the auxkernel at nodes? What if now I directly compute velocity (one of unknowns) from solving the model equations implicitly and simultaneously? Is it still nodal variable? then how does moose implement the quadrature integration? Thank you, I just recall some stuff in my mind.

Best regards,
Qiao

**Reply by GiudGiud** [reply]
~~The quadrature points for nodal variables are the nodes.~~

Nodal variables may be auxiliary or nonlinear.
Quadrature points are actually not at nodes, but some loops in MOOSE over quadrature points can be re-used for nodal values, by just pretending the node is the only Qp
